cpp/src/arrow/flight/sql/odbc/odbc_api.cc

```diff
@@ -121,6 +123,107 @@ SQLRETURN SQLFreeHandle(SQLSMALLINT type, SQLHANDLE handle) {
   return SQL_ERROR;
 }
 
+SQLRETURN SQLGetDiagField(SQLSMALLINT handleType, SQLHANDLE handle, SQLSMALLINT recNumber,
+                          SQLSMALLINT diagIdentifier, SQLPOINTER diagInfoPtr,
+                          SQLSMALLINT bufferLength, SQLSMALLINT* stringLengthPtr) {
+  using driver::odbcabstraction::Diagnostics;
+  using ODBC::GetStringAttribute;
+  using ODBC::ODBCConnection;
+  using ODBC::ODBCEnvironment;
+
+  if (!handle) {
+    return SQL_INVALID_HANDLE;
+  }
+
+  if (!diagInfoPtr) {
+    return SQL_ERROR;
+  }
+
+  // Set character type to be Unicode by defualt (not Ansi)
+  bool isUnicode = true;
+  Diagnostics* diagnostics = nullptr;
+
+  switch (handleType) {
+    case SQL_HANDLE_ENV: {
+      ODBCEnvironment* environment = reinterpret_cast<ODBCEnvironment*>(handle);
+      diagnostics = &environment->GetDiagnostics();
+      break;
+    }
+
+    case SQL_HANDLE_DBC: {
+      ODBCConnection* environment = reinterpret_cast<ODBCConnection*>(handle);
+      diagnostics = &environment->GetDiagnostics();
+      break;
+    }
+
+    default:
+      return SQL_ERROR;
+  }
+
+  if (!diagnostics) {
+    return SQL_INVALID_HANDLE;
```

Comment:
```suggestion
    return SQL_ERROR;
```
I think it should return error here because the handle was not null, and an issue occurred

Reply:
Changed to return SQL_ERROR when returned diagnostics is null.